python/nav/web/auth.py

```diff
@@ -24,57 +24,107 @@
-                    ext_sync='ldap'
-                )
-                account.set_password(password)
-                account.save()
-                # We're authenticated now
-                auth = True
-
-    if account and account.locked:
+        return None
+
+    # Bail out! Potentially evil user
+    if account.locked:
         logger.info("Locked user %s tried to log in", account.login)
+        # Needs auditlog
+        return False
 
-    if (account and
-            account.ext_sync == 'ldap' and
-            ldapauth.available and
-            not auth and
-            not account.locked):
-        try:
-            auth = ldapauth.authenticate(username, password)
-        except ldapauth.NoAnswerError:
-            # Fallback to stored password if ldap is unavailable
-            auth = False
-        else:
-            if auth:
-                account.set_password(password)
-                account.save()
-            else:
-                return
-
-    if account and not auth:
-        auth = account.check_password(password)
-
-    if auth and account:
+    if account.check_password(password):
         return account
-    else:
+
+    # Password was incorrect
+    return None
+
+
+def authenticate_ldap(username=None, password=None):
+    """
+    Authenticate username and password against LDAP, if available
+
+    Returns account object if user was authenticated, False if the user is
+    invlid or otherwise is not allowed to log in, else None.
+    """
+    if not username and not password:
         return None
+
+    if not ldapauth.available:
+        return None
+
+    try:
+        ldapuser = ldapauth.authenticate(username, password)
+    except ldapauth.NoAnswerError:
+        # LDAP unreachable, fallback
+        return None
+
+    if ldapuser is False:
+        # This user does not exist in LDAP, abort!
+        return False
+
+    # From this point on we have an authenticated LDAPUser
+
+    try:
+        account = Account.objects.get(login__iexact=username)
+    except Account.DoesNotExist:
+        # Store the ldapuser in the database and return the new account
+        account = Account(
+            login=user.username,
+            name=user.get_real_name(),
+            ext_sync='ldap'
+        )
+        account.set_password(password)
+        account.save()
+        logger.info("Created user %s from LDAP", account.login)
+        return account
+
+    # From this point on we have an existing Account
+
+    # Bail out! Potentially evil user
+    if account.locked:
+        logger.info("Locked ldap user %s tried to log in", account.login)
+        # Needs auditlog
+        return False
+
+    # Sync password from ldap to local db
+    if not account.check_password(password):
+        account.set_password(password)
+        logger.info("Synced user %s's password from LDAP", account.login)
+        account.save()
+
+    return account
+
+
+def authenticate(username, password):
```

Comment:
As "Uncle Bob" put it, code should read like a book. I'd much prefer the meat of this module, i.e. the `authenticate()` function  should be first. Then, if I need to consult what the callouts do, I should find the called functions if I read further down.